docs/pages/principles.rst

```diff
@@ -0,0 +1,89 @@
+Principles
+==========
+
+This section describes a few elements which are at the core of ``Collection``, 
+and provides guidance on how to get the most out of the package.
+
+Immutability
+------------
+
+Most :ref:`operations <Methods (operations)>` return a new collection object
+rather than modifying the original one. This makes your code easier to reason
+about and more testable.
+
+A few notable exceptions to this are methods which return scalar values like ``count``;
+you can read more about them in the :ref:`operations background <Background>` section. 
+
+Laziness
+--------
+
+Behaviour
+~~~~~~~~~
+
+``Collection`` leverages PHP ``Generators`` and ``Closures`` to allow working with
+large data sets while using as little memory as possible. Operations benefit from
+`lazy evaluation`_, which means that even using multiple methods in succession
+will not have much impact on the program's runtime until the collection is iterated on, or
+a dedicated operation such as ``all`` or ``squash`` is used.
+
+This feature allows us to work even with potentially infinite data sets or streams of data.
+
+.. code-block:: php
+
+    $even = static fn(int $item): bool => $item % 2 === 0;
+    $square = static fn(int $item): int => $item * $item;
+
+    $collection = Collection::fromIterable(range(0, 5))
+        ->filter($even)
+        ->map($square); // $collection is still unchanged up to this point
+
+    // We can use `all` to get the results as an array
+    var_dump($collection->all()); // [0, 4, 16]
+
+    // Even better, we can iterate over the collection directly
+    foreach ($collection as $item) {
+        var_dump($item); // will print 0, 4, 16
+    }
+
+Internals
+~~~~~~~~~
+
+The ``Collection`` object is assisted by a couple powerful components:
+
+`ClosureIterator`_ - allows iterating over the collection object by creating 
+a new PHP Generator each time iteration is started.
+
+`AbstractOperation`_ and the `Operation Interface`_ - provide the blueprint
+for collection operations, which are pure functions defined as final classes
+with the `invoke`_ PHP magic method. Operations return a ``Closure`` which itself
+returns an ``Iterator``; they can thus be used inside ``ClosureIterator`` to 
+create new generators when needed.
+
+Side-Effects
+~~~~~~~~~~~~
+
+Sometimes you might want your ``Collection`` object to behave eagerly.
+This is a legitimate need when you want to trigger side-effects, like persisting
+entities to a database or sending a batch of emails.
```

Comment:
haha ! People I work with are using `Collection` for doing things like that. As I'm doing the code reviews most of the time, I always ask them to use a regular `foreach` loop instead.
Why?
Because I ultimately think that using `Collection` for doing tasks with side-effects is not the proper way to use this tool.
According to me, if you use `Collection` without using it's return value, then you're using it wrong.
I'm definitely aware that technically it will work, but semantically, it's worse than using a plain old `foreach` (_or alike_) loop.

Reply:
hmm I get what you mean, I guess I'm just not a big fan of loops in general 😅 .

I see `Collection::apply` as a replacement for a separate `foreach` loop (ofc you have to use `squash` as well), because to me doing `apply` + `foreach` seems redundant.

In Rust, iterators have a [for_each](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.for_each) method which basically acts like `apply` but is _not_ lazy, which means it can be a substitute for a loop when you want to do some side effects, especially at the end of a chain of other transformations: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3bcb45d0c7f99d3a91e98e65cf87c371

Reply:
This being said, I'll add a note that normal iteration via a `foreach` loop should be preferred for triggering side-effects

Reply:
Ok I see, thanks for the Rust example.

Maybe add a note in this direction:

`Collection` is a helper for making transformations to input data. Despite the fact that you can technically do side-effects in some operations(_due to the fact that you can pass a custom Closure that can alter variables out of its scope_) , it's better to avoid using the library for such endeavor, but rather to get the return value of a transformation.

Reply:
@drupol thank you for the suggestion, I've updated the copy now.

I'm still conflicted about the `apply` operation: maybe I'm looking at it the wrong way, but what is it's purpose if not to replace a `foreach` loop? I see it as the equivalent of a loop and similar to Laravel Collection's [each](https://laravel.com/docs/8.x/collections#method-each) method.

Reply:
There's no conflict, no worries. `Apply` has been made for that, but I don't advise to use it at all, that's it :)